[JENKINS-24673] SimpleBuildWrapper

jglick opened the pull request:
[JENKINS-24673](https://issues.jenkins-ci.org/browse/JENKINS-24673)

Should provide an easy compatibility transition for many common `BuildWrapper`s within a workflow. Does not currently support the more exotic capabilities of `BuildWrapper`; only the basic set up/tear down lifecycle and overriding of environment variables. Generalizes https://github.com/jenkinsci/credentials-binding-plugin/pull/3.
- [X] define basic interface
- [x] create corresponding `Step`: https://github.com/jenkinsci/workflow-plugin/issues/37
- [x] prototype integration in `config-file-provider`: https://github.com/jenkinsci/config-file-provider-plugin/pull/9
- [x] prototype integration in `xvnc`: https://github.com/jenkinsci/xvnc-plugin/pull/5
- [ ] prototype integration in `ssh-agent`
- [ ] prototype integration in `android-emulator`

[Known build wrappers](https://github.com/search?type=Code&q=user%3Ajenkinsci+%22extends+BuildWrapper%22) (until [INFRA-180](https://issues.jenkins-ci.org/browse/INFRA-180) is fixed)

@reviewbybees (including downstream PRs)

jglick committed:
Initial draft of SimpleBuildWrapper.

jglick committed:
Compilability fix.
For some reason (probably a mistake), BuildWrapper.Environment is nonstatic.

jglick committed:
tearDown should be allowed to throw exceptions.

jglick committed:
Merge branch 'master' into SimpleBuildWrapper-JENKINS-24673

jglick committed:
API refinements and documentation.

jglick committed:
Typo.

rsandell commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -0,0 +1,211 @@
+         */
+        public void env(String key, String value) {
+            if (env.put(key, value) != null) {
+                throw new IllegalStateException("just one binding for " + key);
```
What is the purpose of throwing this? The value seems to have been replaced after the throw, so a caller could just catch the exception to have the value overwritten anyways.

jglick replied:
True, but at least you would notice that you had a bug that ought to be fixed. Can change it to check before the `put` if you care.

jglick committed:
Merge branch 'master' into SimpleBuildWrapper-JENKINS-24673

jglick committed:
Better to throw the ISE before, not after, overwriting the environment variable.

rsandell commented:
+1

jglick commented:
@kohsuke review would be welcome.

kohsuke commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -0,0 +1,212 @@
+     */
+    public static final class Context {
+        private Disposer disposer;
+        private final Map<String,String> env = new HashMap<String,String>();
```
If this is to store environment variables, I'd encourage the use of `EnvVars` class for proper name comparison semantics.

jglick replied:
That is actually what I had originally, but it seems that `Map<String,String>` works better here. See `EnvironmentWrapper.buildEnvVars`, calling `EnvVars.putAll`, which takes a `Map`. More to the point, see https://github.com/jenkinsci/workflow-plugin/pull/37/files#diff-08a6c4ff385f29bd9c89e3395348094cR81 where `overrideAll` is called, which also takes a `Map` and indeed treats keys specially if using `+` syntax.

The one thing that would sway me is if `EnvVars` were able to carry its own set of sensitive keys, as mentioned below in the comment in `makeSensitiveBuildVariables`. This seems like a more complicated change since it would involve deprecating some things in `AbstractBuild`.

jglick replied:
> if `EnvVars` were able to carry its own set of sensitive keys

Though even if it did, the signature of `Context` could easily be evolved to let that be used here.

kohsuke commented:
Other than one minor comment, it looks good to me.

jglick committed:
Javadoc improvements.

jglick committed:
No need to check for c.env == null any more.

jglick committed:
Added some tests for SimpleBuildWrapper used from a freestyle project as a BuildWrapper.
This demonstrated that we actually need to call EnvVars.overrideAll explicitly (as CoreWrapperStep already did).

jglick committed:
Also verifying that initialEnvironment somehow works.

jglick commented:
@reviewbybees recent corrections.

kohsuke commented:
This looks good to me.

jtnord commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -0,0 +1,224 @@
+
+    @Override public final Environment setUp(AbstractBuild build, final Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
+        if (runPreCheckout()) {
+            return new Environment() {};
```
this inner class will have a reference to this BuildWrapper which is never used.  
May be better to have a new EmptyEnvironment class.

jglick replied:
The reference is harmless however, because in this context the `Environment` is not serialized, and the `BuildWrapper` will last at least as long (the `buildEnvironments` are cleared in `AbstractBuildExecution.cleanUp`).

jtnord commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -0,0 +1,224 @@
+        }
+    }
+
+    private class EnvironmentWrapper extends Environment {
```
first glance says this should be a static class as reference to the enclosing Object is not used.

jglick replied:
Indeed it is not, but `BuildWrapper.Environment` is nonstatic (probably an API mistake).

jglick commented:
@jtnord agreed?

jtnord commented:
@jglick this time only :)
:+1:
